src/Common/src/System/Runtime/InteropServices/StringBuffer.cs

```diff
@@ -0,0 +1,385 @@
+        {
+            get { return _length; }
+            set
+            {
+                // Null terminate
+                EnsureCapacity(value + 1);
+                CharPointer[value] = '\0';
+
+                _length = value;
+            }
+        }
+
+        /// <summary>
+        /// For use when the native api null terminates but doesn't return a length.
+        /// If no null is found, the length will not be changed.
+        /// </summary>
+        public unsafe void SetLengthToFirstNull()
+        {
+            char* buffer = CharPointer;
+            ulong capacity = CharCapacity;
+            for (ulong i = 0; i < capacity; i++)
+            {
+                if (buffer[i] == '\0')
+                {
+                    _length = i;
+                    break;
+                }
+            }
+        }
+
+        internal unsafe char* CharPointer
+        {
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            get
+            {
+                return (char*)VoidPointer;
+            }
+        }
+
+        /// <summary>
+        /// Returns true if the buffer starts with the given string.
+        /// </summary>
+        public bool StartsWithOrdinal(string value)
+        {
+            if (value == null) throw new ArgumentNullException("value");
+            if (_length < (ulong)value.Length) return false;
+            return SubStringEquals(value, startIndex: 0, count: value.Length);
+        }
+
+        /// <summary>
+        /// Returns true if the specified StringBuffer substring equals the given value.
+        /// </summary>
+        /// <param name="value">The value to compare against the specified substring.</param>
+        /// <param name="startIndex">Start index of the sub string.</param>
+        /// <param name="count">Length of the substring, or -1 to check all remaining.</param>
+        /// <exception cref="ArgumentOutOfRangeException">
+        /// Thrown if <paramref name="startIndex"/> or <paramref name="count"/> are outside the range
+        /// of the buffer's length.
+        /// </exception>
+        public unsafe bool SubStringEquals(string value, ulong startIndex = 0, int count = -1)
+        {
+            if (value == null) return false;
+            if (count < -1) throw new ArgumentOutOfRangeException("count");
+            ulong realCount = count == -1 ? _length - startIndex : (ulong)count;
+            if (startIndex + realCount > _length) throw new ArgumentOutOfRangeException("count");
+
+            int length = value.Length;
+
+            // Check the substring length against the input length
+            if (realCount != (ulong)length) return false;
+
+            fixed (char* valueStart = value)
+            {
+                char* bufferStart = CharPointer + startIndex;
+                for (int i = 0; i < length; i++)
+                {
+                    if (*bufferStart++ != valueStart[i]) return false;
+                }
```

Comment:
You could rewrite this as:
```
while (length--)
{
    if (*bufferStart++ != **valueStart++)
        return false;
}
```
to avoid needing to do the offset calculation into valueStart, and for consistency between the two sides of the equality check.